[WebGPU] Implement resource deallocation

mwyrzykowski opened the pull request:
#### 16f668371d1ab49a5cb205a706b8ca9e370d18f4
<pre>
[WebGPU] Implement resource deallocation
<a href="https://bugs.webkit.org/show_bug.cgi?id=250865">https://bugs.webkit.org/show_bug.cgi?id=250865</a>
&lt;radar://103963935&gt;

Reviewed by NOBODY (OOPS!).

Resources were not getting removed from the GPUProcess
when they were no longer needed in the WebKit process.

Fix this by adding messages the WebKit process can send to
the GPU process to remove resources.

Also add a notification facility to show the number of
active objects in the WebGPUObjectHeap.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSurfaceImpl.cpp:
(PAL::WebGPU::SurfaceImpl::destroy):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSwapChainImpl.cpp:
(PAL::WebGPU::SwapChainImpl::~SwapChainImpl):
(PAL::WebGPU::SwapChainImpl::destroy):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::getBindGroupLayout):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:
(WebKit::RemoteAdapter::~RemoteAdapter):
(WebKit::RemoteAdapter::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.cpp:
(WebKit::RemoteBindGroup::~RemoteBindGroup):
(WebKit::RemoteBindGroup::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroupLayout.cpp:
(WebKit::RemoteBindGroupLayout::~RemoteBindGroupLayout):
(WebKit::RemoteBindGroupLayout::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroupLayout.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroupLayout.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::~RemoteBuffer):
(WebKit::RemoteBuffer::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandBuffer.cpp:
(WebKit::RemoteCommandBuffer::~RemoteCommandBuffer):
(WebKit::RemoteCommandBuffer::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandBuffer.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandBuffer.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.cpp:
(WebKit::RemoteCommandEncoder::~RemoteCommandEncoder):
(WebKit::RemoteCommandEncoder::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePassEncoder.cpp:
(WebKit::RemoteComputePassEncoder::~RemoteComputePassEncoder):
(WebKit::RemoteComputePassEncoder::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePassEncoder.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePassEncoder.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePipeline.cpp:
(WebKit::RemoteComputePipeline::~RemoteComputePipeline):
(WebKit::RemoteComputePipeline::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePipeline.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePipeline.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::~RemoteDevice):
(WebKit::RemoteDevice::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteExternalTexture.cpp:
(WebKit::RemoteExternalTexture::~RemoteExternalTexture):
(WebKit::RemoteExternalTexture::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteExternalTexture.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteExternalTexture.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePipelineLayout.cpp:
(WebKit::RemotePipelineLayout::~RemotePipelineLayout):
(WebKit::RemotePipelineLayout::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePipelineLayout.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePipelineLayout.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQuerySet.cpp:
(WebKit::RemoteQuerySet::~RemoteQuerySet):
(WebKit::RemoteQuerySet::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.cpp:
(WebKit::RemoteQueue::~RemoteQueue):
(WebKit::RemoteQueue::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundle.cpp:
(WebKit::RemoteRenderBundle::~RemoteRenderBundle):
(WebKit::RemoteRenderBundle::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundle.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundle.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundleEncoder.cpp:
(WebKit::RemoteRenderBundleEncoder::~RemoteRenderBundleEncoder):
(WebKit::RemoteRenderBundleEncoder::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundleEncoder.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundleEncoder.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPassEncoder.cpp:
(WebKit::RemoteRenderPassEncoder::~RemoteRenderPassEncoder):
(WebKit::RemoteRenderPassEncoder::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPassEncoder.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPassEncoder.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPipeline.cpp:
(WebKit::RemoteRenderPipeline::~RemoteRenderPipeline):
(WebKit::RemoteRenderPipeline::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPipeline.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPipeline.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSampler.cpp:
(WebKit::RemoteSampler::~RemoteSampler):
(WebKit::RemoteSampler::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSampler.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSampler.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteShaderModule.cpp:
(WebKit::RemoteShaderModule::~RemoteShaderModule):
(WebKit::RemoteShaderModule::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteShaderModule.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteShaderModule.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSurface.cpp:
(WebKit::RemoteSurface::~RemoteSurface):
(WebKit::RemoteSurface::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSwapChain.cpp:
(WebKit::RemoteSwapChain::~RemoteSwapChain):
(WebKit::RemoteSwapChain::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTexture.cpp:
(WebKit::RemoteTexture::~RemoteTexture):
(WebKit::RemoteTexture::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTextureView.cpp:
(WebKit::RemoteTextureView::~RemoteTextureView):
(WebKit::RemoteTextureView::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTextureView.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTextureView.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.cpp:
(WebKit::WebGPU::ObjectHeap::typeOfObject):
(WebKit::WebGPU::ObjectHeap::printObjectHeap):
(WebKit::WebGPU::ObjectHeap::ObjectHeap):
(WebKit::WebGPU::ObjectHeap::removeObject):
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp:
(WebKit::WebGPU::RemoteAdapterProxy::~RemoteAdapterProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBindGroupLayoutProxy.cpp:
(WebKit::WebGPU::RemoteBindGroupLayoutProxy::~RemoteBindGroupLayoutProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBindGroupLayoutProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBindGroupProxy.cpp:
(WebKit::WebGPU::RemoteBindGroupProxy::~RemoteBindGroupProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBindGroupProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::~RemoteBufferProxy):
(WebKit::WebGPU::RemoteBufferProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCommandBufferProxy.cpp:
(WebKit::WebGPU::RemoteCommandBufferProxy::~RemoteCommandBufferProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCommandBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCommandEncoderProxy.cpp:
(WebKit::WebGPU::RemoteCommandEncoderProxy::~RemoteCommandEncoderProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCommandEncoderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePassEncoderProxy.cpp:
(WebKit::WebGPU::RemoteComputePassEncoderProxy::~RemoteComputePassEncoderProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePassEncoderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePipelineProxy.cpp:
(WebKit::WebGPU::RemoteComputePipelineProxy::~RemoteComputePipelineProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePipelineProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::~RemoteDeviceProxy):
(WebKit::WebGPU::RemoteDeviceProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteExternalTextureProxy.cpp:
(WebKit::WebGPU::RemoteExternalTextureProxy::~RemoteExternalTextureProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteExternalTextureProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::~RemoteGPUProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemotePipelineLayoutProxy.cpp:
(WebKit::WebGPU::RemotePipelineLayoutProxy::~RemotePipelineLayoutProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemotePipelineLayoutProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQuerySetProxy.cpp:
(WebKit::WebGPU::RemoteQuerySetProxy::~RemoteQuerySetProxy):
(WebKit::WebGPU::RemoteQuerySetProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQuerySetProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
(WebKit::WebGPU::RemoteQueueProxy::~RemoteQueueProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderBundleEncoderProxy.cpp:
(WebKit::WebGPU::RemoteRenderBundleEncoderProxy::~RemoteRenderBundleEncoderProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderBundleEncoderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderBundleProxy.cpp:
(WebKit::WebGPU::RemoteRenderBundleProxy::~RemoteRenderBundleProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderBundleProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPassEncoderProxy.cpp:
(WebKit::WebGPU::RemoteRenderPassEncoderProxy::~RemoteRenderPassEncoderProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPassEncoderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPipelineProxy.cpp:
(WebKit::WebGPU::RemoteRenderPipelineProxy::~RemoteRenderPipelineProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPipelineProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSamplerProxy.cpp:
(WebKit::WebGPU::RemoteSamplerProxy::~RemoteSamplerProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSamplerProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.cpp:
(WebKit::WebGPU::RemoteShaderModuleProxy::~RemoteShaderModuleProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSurfaceProxy.cpp:
(WebKit::WebGPU::RemoteSurfaceProxy::~RemoteSurfaceProxy):
(WebKit::WebGPU::RemoteSurfaceProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSurfaceProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSwapChainProxy.cpp:
(WebKit::WebGPU::RemoteSwapChainProxy::~RemoteSwapChainProxy):
(WebKit::WebGPU::RemoteSwapChainProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSwapChainProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
(WebKit::WebGPU::RemoteTextureProxy::~RemoteTextureProxy):
(WebKit::WebGPU::RemoteTextureProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureViewProxy.cpp:
(WebKit::WebGPU::RemoteTextureViewProxy::~RemoteTextureViewProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureViewProxy.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/16f668371d1ab49a5cb205a706b8ca9e370d18f4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104634 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13708 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37535 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113909 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174124 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108550 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14833 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4639 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96964 "Built successfully") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112849 "Hash 16f66837 for PR 8856 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110397 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11446 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94476 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39009 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/108106 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93302 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26088 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80673 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94593 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7065 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27446 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92526 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4834 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7173 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4021 "Found 2 new test failures: fast/images/avif-heif-container-as-image.html, fast/images/avif-image-document.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30097 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/103463 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13220 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47002 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101212 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8962 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25164 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h:
```diff
@@ -69,21 +69,27 @@ class RemoteAdapterProxy final : public PAL::WebGPU::Adapter {
     template<typename T>
     WARN_UNUSED_RETURN bool send(T&& message)
     {
+        if (m_destroyed)
```
I'm not sure if there is a better / alternate way of achieving this, maybe @kkinnunen-apple knows?

It is to avoid the issue where destroy is sent from, for example, GPUAdapter prior to the destructor of GPUAdapter being called. But then destroy gets sent again in the destructor of GPUAdapter, but there is no receiver for the message so we encounter an IPC error.

The logic doesn't really apply to GPUAdapter but does apply to time limited resources, e.g., GPUCommandEncoder, GPUTextureView, etc

litherum replied:
Can't we avoid this by construction, and say "the only place that sends destroy messages are the destructors of RemoteFooProxy classes?"

litherum commented on Source/WebCore/Modules/WebGPU/GPUBindGroup.cpp:
```diff
@@ -28,6 +28,11 @@
 
+GPUBindGroup::~GPUBindGroup()
+{
+    m_backing->destroy();
```
I think, either:
A) the mapping between WebCore types and WebKit types is one-to-one, in which case the destructor for the webkit type will be called here without us having to do anything in WebCore, or
B) the mapping between them isn't one-to-one, and something else has retained a WebKit type, and that thing will be left with a dangling pointer after this code executes.

So I kind of feel that WebCore should be left alone here, and destruction should be driven by WebKit objects's destructors in the web process.

litherum commented on Source/WebCore/Modules/WebGPU/GPUTexture.cpp:
```diff
@@ -32,6 +32,11 @@
 
+GPUTexture::~GPUTexture()
+{
+    destroy();
```
I actually think this is construing semantics. The API-exposed `destroy()` functions are supposed to mean "the browser's object is alive - and can continue to be interacted with, methods called on it, validation is performed, state tracking is still performed, etc, but there's just no Metal object inside.

That's different than what you're doing here, which is trying to communicate "the browser object doesn't need to exist any more at all" between processes, to avoid leaks. I think we should treat those 2 things as distinct, because I believe they have observably different semantics.

litherum commented on Source/WebCore/Modules/WebGPU/GPUQuerySet.h:
```diff
@@ -47,6 +47,7 @@ class GPUQuerySet : public RefCounted<GPUQuerySet> {
     PAL::WebGPU::QuerySet& backing() { return m_backing; }
     const PAL::WebGPU::QuerySet& backing() const { return m_backing; }
 
+    ~GPUQuerySet();
```
blank line before `private:`, please.

litherum commented on Source/WebCore/Modules/WebGPU/GPUCommandEncoder.cpp:
```diff
@@ -134,7 +134,14 @@ static PAL::WebGPU::CommandBufferDescriptor convertToBacking(const std::optional
 {
-    return GPUCommandBuffer::create(m_backing->finish(convertToBacking(commandBufferDescriptor)));
+    auto result = GPUCommandBuffer::create(m_backing->finish(convertToBacking(commandBufferDescriptor)));
+    m_backing->destroy();
```
Is this right? See my other comment about how the browser objects need to still exist for validation, state tracking, etc.

litherum commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -59,6 +59,13 @@ void GPUQueue::submit(Vector<RefPtr<GPUCommandBuffer>>&& commandBuffers)
+        if (!commandBuffer)
+            continue;
+
+        commandBuffer->destroy();
```
Ditto.

litherum commented on Source/WebCore/Modules/WebGPU/GPUTexture.cpp:
```diff
@@ -61,11 +66,29 @@ static PAL::WebGPU::TextureViewDescriptor convertToBacking(const std::optional<G
+    auto textureView = GPUTextureView::create(m_backing->createView(convertToBacking(textureViewDescriptor)));
+    m_cachedTextureViews.append(std::make_pair(textureViewDescriptor, textureView));
+
+    return textureView;
```
Can we do this in a follow-up patch? Implementing a texture view cache seems different than "no objects are getting destroyed ever"

litherum commented on Source/WebCore/Modules/WebGPU/GPUTexture.h:
```diff
@@ -60,6 +63,7 @@ class GPUTexture : public RefCounted<GPUTexture> {
     }
 
     Ref<PAL::WebGPU::Texture> m_backing;
+    mutable Vector<std::pair<std::optional<GPUTextureViewDescriptor>, Ref<GPUTextureView>>> m_cachedTextureViews;
```
I think this is conceptually backwards from an ownership perspective. Objects can hold strong references to their parent object, not the other way around.

E.g. retaining a device shouldn't retain all objects that have ever been created from that device; instead, if a JS application has created all the objects they'll need, they should be free to drop all references to the Device, but the device can stay alive anyway because it can have child objects that retain it. So, the ownership model is inverted, and branch objects shouldn't retain the leaf objects they've created; instead, the leafs can retain the branches.

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUQueueImpl.h:
```diff
@@ -83,6 +83,7 @@ class QueueImpl final : public Queue {
         const Extent3D& copySize) final;
 
     void setLabelInternal(const String&) final;
+    void destroy() final { }
```
Wait, why does the Impl need destroy() methods? The purpose of the destroy() methods is to remove objects from the WebKit::ObjectHeap, which will cause them to be destroyed; no need for methods on objects to tell them to destroy themselves...

litherum replied:
I'm confused, why do all these objects have destroy() methods if they don't actually do anything?

litherum commented on Source/WebCore/html/canvas/GPUCanvasContext.h:
```diff
@@ -138,6 +138,10 @@ class GPUCanvasContext final : public GPUBasedCanvasRenderingContext {
+    struct GPUTextureHolder {
+        Ref<GPUTexture> texture;
+    };
+    std::unique_ptr<GPUTextureHolder> m_textureHolder;
```
I don't understand. Why do we need a texture holder?

I think this lifetime management of render buffers should be done in a separate patch; it's a distinct problem from "nothing is getting destroyed ever."

Also, I'm making a lot of changes in this area; we should coordinate to make sure we don't step on each others' toes.

litherum commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -548,14 +548,15 @@ static WGPUBindGroupLayoutEntry createEntryFromStructMember(MTLStructMember *str
 #endif // HAVE(METAL_BUFFER_BINDING_REFLECTION)
 
-BindGroupLayout* RenderPipeline::getBindGroupLayout(uint32_t groupIndex)
+Ref<BindGroupLayout> RenderPipeline::getBindGroupLayout(uint32_t groupIndex, bool& createdLayout)
```
I think this is wrong. Getters should return +0 objects, not +1 objects. Ref indicates it's a +1 object. Instead, getters are supposed to return a raw (or weak) pointer (or C++ reference) to an internally-owned object.

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.cpp:
```diff
@@ -44,7 +44,16 @@ RemoteBindGroup::RemoteBindGroup(PAL::WebGPU::BindGroup& bindGroup, WebGPU::Obje
-RemoteBindGroup::~RemoteBindGroup() = default;
+RemoteBindGroup::~RemoteBindGroup()
+{
+    destroy();
```
I'm super confused about this. When we receive the Destroy message, we run destroy(), which removes us from the object heap. That will cause our refcount to reach 0, which will cause us to be destroyed, which will run our destructor.... which will run the destroy() function again???

Why isn't it just:
```
RemoteBindGroup::~RemoteBindGroup() = default;

void RemoteBindGroup::destroy()
{
    m_objectHeap.removeObject(m_identifier);
    // we're dead now, don't access anything! return immediately.
}
```
and that's it?

mwyrzykowski replied:
I think that's right, we shouldn't need two calls to destroy. I'll try with the default destructor approach

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.cpp:
```diff
@@ -53,10 +53,119 @@
+
 ObjectHeap::ObjectHeap()
 {
+#if PLATFORM(COCOA)
```
Can we only do this in debug builds?

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.cpp:
```diff
@@ -53,10 +53,119 @@
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [] {
+        int printObjectHeapToken;
+        notify_register_dispatch("com.apple.WebKit.WebGPU.PrintObjectHeap", &printObjectHeapToken, dispatch_get_main_queue(), ^(int) {
```
We should be able to pass C++ lambdas to functions that accept blocks. That's better in our situation, because we can put stuff in the capture clause.

```
notify_register_dispatch("com.apple.WebKit.WebGPU.PrintObjectHeap", &printObjectHeapToken, dispatch_get_main_queue(), [objectHeap = Ref { *this }] () {
    objectHeap->printObjectHeap();
});
```

(if that doesn't work, we definitely can do `createBlockPtr([...stuff...] () { ... stuff ... }).get()`.)

(This technique may require `#include <wtf/BlockPtr.h>` too, not sure.)

I also think we shouldn't have to deal with all this `call_once` stuff, and just have every object heap that exists register for printing. If you have 2 alive, you probably want both to be dumped, instead of whichever was created second.

mwyrzykowski replied:
Good point I will add the token as a member variable and cancel the registration in the destructor.

I don't think we should capture `*this` in a `Ref<>` because then the `notify_register_dispatch` will retain the object and keep the object alive right? We don't want it retained by `notify_register_dispatch`

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.cpp:
```diff
@@ -53,10 +53,119 @@
 {
+#if PLATFORM(COCOA)
+    static ObjectHeap* currentObjectHeap;
+    currentObjectHeap = this;
```
This may be racey; better to just have all object heaps register for printing

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.cpp:
```diff
@@ -53,10 +53,119 @@
+
 namespace WebKit::WebGPU {
 
+#if PLATFORM(COCOA)
```
There's nothing platform specific about printing a heap; can we keep these outside of the guard, and just guard the `dispatch` stuff that really is platform specific?

litherum reviewed (changes requested):
I think this patch should be made much simpler, keeping all the logic inside WebKit/, and have this just hook up RemoteFooProxy objects' destruction to RemoteFoo objects' removal from the object heap. No need for anything in PAL or WebCore to even be aware this is happening.

litherum reviewed (changes requested):
I think this patch should be made much simpler, keeping all the logic inside WebKit/, and have this just hook up RemoteFooProxy objects' destruction to RemoteFoo objects' removal from the object heap. No need for anything in PAL or WebCore to even be aware this is happening.

mwyrzykowski committed:
[WebGPU] Implement resource deallocation
https://bugs.webkit.org/show_bug.cgi?id=250865
<radar://103963935>

Reviewed by NOBODY (OOPS!).

Resources were not getting removed from the GPUProcess
when they were no longer needed in the WebKit process.

Fix this by adding messages the WebKit process can send to
the GPU process to remove resources.

Also add a notification facility to show the number of
active objects in the WebGPUObjectHeap.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSurfaceImpl.cpp:
(PAL::WebGPU::SurfaceImpl::destroy):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSwapChainImpl.cpp:
(PAL::WebGPU::SwapChainImpl::~SwapChainImpl):
(PAL::WebGPU::SwapChainImpl::destroy):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::getBindGroupLayout):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.cpp:
(WebKit::RemoteAdapter::~RemoteAdapter):
(WebKit::RemoteAdapter::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteAdapter.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.cpp:
(WebKit::RemoteBindGroup::~RemoteBindGroup):
(WebKit::RemoteBindGroup::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroup.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroupLayout.cpp:
(WebKit::RemoteBindGroupLayout::~RemoteBindGroupLayout):
(WebKit::RemoteBindGroupLayout::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroupLayout.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBindGroupLayout.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
(WebKit::RemoteBuffer::~RemoteBuffer):
(WebKit::RemoteBuffer::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandBuffer.cpp:
(WebKit::RemoteCommandBuffer::~RemoteCommandBuffer):
(WebKit::RemoteCommandBuffer::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandBuffer.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandBuffer.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.cpp:
(WebKit::RemoteCommandEncoder::~RemoteCommandEncoder):
(WebKit::RemoteCommandEncoder::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePassEncoder.cpp:
(WebKit::RemoteComputePassEncoder::~RemoteComputePassEncoder):
(WebKit::RemoteComputePassEncoder::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePassEncoder.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePassEncoder.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePipeline.cpp:
(WebKit::RemoteComputePipeline::~RemoteComputePipeline):
(WebKit::RemoteComputePipeline::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePipeline.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePipeline.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::~RemoteDevice):
(WebKit::RemoteDevice::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteExternalTexture.cpp:
(WebKit::RemoteExternalTexture::~RemoteExternalTexture):
(WebKit::RemoteExternalTexture::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteExternalTexture.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteExternalTexture.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.cpp:
(WebKit::RemoteGPU::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteGPU.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePipelineLayout.cpp:
(WebKit::RemotePipelineLayout::~RemotePipelineLayout):
(WebKit::RemotePipelineLayout::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePipelineLayout.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemotePipelineLayout.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQuerySet.cpp:
(WebKit::RemoteQuerySet::~RemoteQuerySet):
(WebKit::RemoteQuerySet::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.cpp:
(WebKit::RemoteQueue::~RemoteQueue):
(WebKit::RemoteQueue::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundle.cpp:
(WebKit::RemoteRenderBundle::~RemoteRenderBundle):
(WebKit::RemoteRenderBundle::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundle.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundle.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundleEncoder.cpp:
(WebKit::RemoteRenderBundleEncoder::~RemoteRenderBundleEncoder):
(WebKit::RemoteRenderBundleEncoder::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundleEncoder.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderBundleEncoder.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPassEncoder.cpp:
(WebKit::RemoteRenderPassEncoder::~RemoteRenderPassEncoder):
(WebKit::RemoteRenderPassEncoder::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPassEncoder.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPassEncoder.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPipeline.cpp:
(WebKit::RemoteRenderPipeline::~RemoteRenderPipeline):
(WebKit::RemoteRenderPipeline::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPipeline.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPipeline.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSampler.cpp:
(WebKit::RemoteSampler::~RemoteSampler):
(WebKit::RemoteSampler::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSampler.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSampler.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteShaderModule.cpp:
(WebKit::RemoteShaderModule::~RemoteShaderModule):
(WebKit::RemoteShaderModule::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteShaderModule.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteShaderModule.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSurface.cpp:
(WebKit::RemoteSurface::~RemoteSurface):
(WebKit::RemoteSurface::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteSwapChain.cpp:
(WebKit::RemoteSwapChain::~RemoteSwapChain):
(WebKit::RemoteSwapChain::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTexture.cpp:
(WebKit::RemoteTexture::~RemoteTexture):
(WebKit::RemoteTexture::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTextureView.cpp:
(WebKit::RemoteTextureView::~RemoteTextureView):
(WebKit::RemoteTextureView::destroy):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTextureView.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteTextureView.messages.in:
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.cpp:
(WebKit::WebGPU::ObjectHeap::typeOfObject):
(WebKit::WebGPU::ObjectHeap::printObjectHeap):
(WebKit::WebGPU::ObjectHeap::ObjectHeap):
(WebKit::WebGPU::ObjectHeap::removeObject):
* Source/WebKit/GPUProcess/graphics/WebGPU/WebGPUObjectHeap.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.cpp:
(WebKit::WebGPU::RemoteAdapterProxy::~RemoteAdapterProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteAdapterProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBindGroupLayoutProxy.cpp:
(WebKit::WebGPU::RemoteBindGroupLayoutProxy::~RemoteBindGroupLayoutProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBindGroupLayoutProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBindGroupProxy.cpp:
(WebKit::WebGPU::RemoteBindGroupProxy::~RemoteBindGroupProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBindGroupProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.cpp:
(WebKit::WebGPU::RemoteBufferProxy::~RemoteBufferProxy):
(WebKit::WebGPU::RemoteBufferProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCommandBufferProxy.cpp:
(WebKit::WebGPU::RemoteCommandBufferProxy::~RemoteCommandBufferProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCommandBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCommandEncoderProxy.cpp:
(WebKit::WebGPU::RemoteCommandEncoderProxy::~RemoteCommandEncoderProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteCommandEncoderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePassEncoderProxy.cpp:
(WebKit::WebGPU::RemoteComputePassEncoderProxy::~RemoteComputePassEncoderProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePassEncoderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePipelineProxy.cpp:
(WebKit::WebGPU::RemoteComputePipelineProxy::~RemoteComputePipelineProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePipelineProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::~RemoteDeviceProxy):
(WebKit::WebGPU::RemoteDeviceProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteExternalTextureProxy.cpp:
(WebKit::WebGPU::RemoteExternalTextureProxy::~RemoteExternalTextureProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteExternalTextureProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp:
(WebKit::RemoteGPUProxy::~RemoteGPUProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemotePipelineLayoutProxy.cpp:
(WebKit::WebGPU::RemotePipelineLayoutProxy::~RemotePipelineLayoutProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemotePipelineLayoutProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQuerySetProxy.cpp:
(WebKit::WebGPU::RemoteQuerySetProxy::~RemoteQuerySetProxy):
(WebKit::WebGPU::RemoteQuerySetProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQuerySetProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.cpp:
(WebKit::WebGPU::RemoteQueueProxy::~RemoteQueueProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteQueueProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderBundleEncoderProxy.cpp:
(WebKit::WebGPU::RemoteRenderBundleEncoderProxy::~RemoteRenderBundleEncoderProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderBundleEncoderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderBundleProxy.cpp:
(WebKit::WebGPU::RemoteRenderBundleProxy::~RemoteRenderBundleProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderBundleProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPassEncoderProxy.cpp:
(WebKit::WebGPU::RemoteRenderPassEncoderProxy::~RemoteRenderPassEncoderProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPassEncoderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPipelineProxy.cpp:
(WebKit::WebGPU::RemoteRenderPipelineProxy::~RemoteRenderPipelineProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPipelineProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSamplerProxy.cpp:
(WebKit::WebGPU::RemoteSamplerProxy::~RemoteSamplerProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSamplerProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.cpp:
(WebKit::WebGPU::RemoteShaderModuleProxy::~RemoteShaderModuleProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSurfaceProxy.cpp:
(WebKit::WebGPU::RemoteSurfaceProxy::~RemoteSurfaceProxy):
(WebKit::WebGPU::RemoteSurfaceProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSurfaceProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSwapChainProxy.cpp:
(WebKit::WebGPU::RemoteSwapChainProxy::~RemoteSwapChainProxy):
(WebKit::WebGPU::RemoteSwapChainProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteSwapChainProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.cpp:
(WebKit::WebGPU::RemoteTextureProxy::~RemoteTextureProxy):
(WebKit::WebGPU::RemoteTextureProxy::destroy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureViewProxy.cpp:
(WebKit::WebGPU::RemoteTextureViewProxy::~RemoteTextureViewProxy):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteTextureViewProxy.h:

mwyrzykowski commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUSurfaceImpl.cpp:
```diff
@@ -49,7 +49,11 @@ SurfaceImpl::~SurfaceImpl()
 
 void SurfaceImpl::destroy()
 {
+    if (!m_backing)
```
these changes are from https://github.com/WebKit/WebKit/pull/8941

which is on the merge-queue so the WebCore changes will go away after I merge that and rebase

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.cpp:
```diff
@@ -96,7 +99,8 @@ void RemoteBuffer::unmap(Vector<uint8_t>&& data)
 {
-    m_backing->destroy();
+    if (m_objectHeap.removeObject(m_identifier))
+        m_backing->destroy();
```
Won't this happen automatically?

mwyrzykowski replied:
because m_backing's destructor should run? Buffer::Buffer's destructor doesn't call destroy, maybe that is fine, but who calls Buffer::destroy() then?

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteCommandEncoder.cpp:
```diff
@@ -48,7 +48,15 @@ RemoteCommandEncoder::RemoteCommandEncoder(PAL::WebGPU::CommandEncoder& commandE
 }
 
-RemoteCommandEncoder::~RemoteCommandEncoder() = default;
+RemoteCommandEncoder::~RemoteCommandEncoder()
+{
+    destroy();
+}
+
+void RemoteCommandEncoder::destroy()
+{
+    m_objectHeap.removeObject(m_identifier);
+}
```
Still confused about this. The destroy() message will be received, then we'll remove the object from the object heap, then that will cause the object's destructor to be called, which will cause destroy() again?

mwyrzykowski replied:
Oh are the Remote* objects only stored in the object heap? Maybe calling destroy from the destructor is not needed, I will try to remove those calls

litherum commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -107,7 +110,8 @@ RemoteQueue& RemoteDevice::queue()
 void RemoteDevice::destroy()
 {
-    m_backing->destroy();
+    if (m_objectHeap.removeObject(m_identifier))
```
A perfect opportunity for a Blade Runner quote!

"Wake up! Time to die."
&nbsp;- Leon
